src/FirebaseAuth.js

```diff
@@ -222,24 +222,27 @@
      */
     _routerMethodOnAuthPromise: function(rejectIfAuthDataIsNull) {
       var ref = this._ref;
-      var deferred = this._q.defer();
 
-      function callback(authData) {
-        if (authData !== null) {
-          deferred.resolve(authData);
-        } else if (rejectIfAuthDataIsNull) {
-          deferred.reject("AUTH_REQUIRED");
-        } else {
-          deferred.resolve(null);
+      return this._utils.promise(function(resolve,reject){
+        function callback(authData) {
+          // Turn off this onAuth() callback since we just needed to get the authentication data once.
+          ref.offAuth(callback);
+
+          if (authData !== null) {
+            resolve(authData);
+            return;
```

Comment:
Hmmm, not quite what I was envisioning. I'd just do `return resolve(authData)` and `return reject("AUTH_REQUIRED")`, etc. No need for it to be on a separate line. And then you can enable the chaining as you mentioned.